File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    class Word {
        int depth;
        String word;
        
        Word(int depth, String word) {
            this.depth = depth;
            this.word = word;
        }
    }
    
    public boolean isConvertable(String a, String b) {
        int cnt = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) cnt++;
        }
        if (cnt == 1) return true;
        else return false;
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        boolean[] visited = new boolean[words.length];
        
        Queue<Word> q = new LinkedList<>();
        q.add(new Word(0, begin));
        
        while (!q.isEmpty()) {
            Word cur = q.poll();
            
            for (int i = 0; i < words.length; i++) {
                if (!visited[i] && isConvertable(cur.word, words[i])) {
                    if (words[i].equals(target)) return cur.depth + 1;
                    visited[i] = true;
                    q.add(new Word(cur.depth + 1, words[i]));
                }
            }
        }
        
        return 0;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <bits/stdc++.h>
#include <vector>

using namespace std;

bool visited[51];

bool isConvertable(string a, string b) {
    int cnt = 0;
    for (int i = 0; i < a.length(); i++) {
        if (a[i] != b[i]) cnt++;
        if (cnt > 1) return false;
    }
    if (cnt == 1) return true;
    else return false;
}

int solution(string begin, string target, vector<string> words) {
    
    queue<pair<int, string>> q;
    q.push({0, begin});
    
    while (!q.empty()) {
        int curDepth = q.front().first;
        string cur = q.front().second;
        q.pop();
        
        for (int i = 0; i < words.size(); i++) {
            if (!visited[i] && isConvertable(cur, words[i])) {
                if (words[i] == target) return curDepth + 1;
                visited[i] = true;
                q.push({curDepth + 1, words[i]});
            }
        }
    }
    
    return 0;
}